### PR TITLE
[rootless] Remove macro support for non-literals, add comment

### PR DESCRIPTION
**Pull request by leptos-null**
What does this implement/fix? Explain your changes.
---------------------------------------------------

Adds a definition for `THEOS_PACKAGE_INSTALL_PREFIX` to empty string if it is not already defined (for example, when importing in Xcode).

Removes `ROOT_PATH_NS_VAR` and `ROOT_PATH_VAR`. These were troublesome macros, because it was difficult to manage the returned memory.

For example, the value provided by `ROOT_PATH_VAR` currently is undefined (given that the implementation uses `{}` which results in the variable going out of scope, allowing the compiler to re-claim the memory). @opa334 [suggested](https://github.com/theos/headers/pull/81#issuecomment-1613650204) in #81 to use [`alloca`](https://stackoverflow.com/a/2318448) to achieve the same goal. Even if we used [`__builtin_alloca`](https://clang.llvm.org/docs/LanguageExtensions.html#builtin-alloca), it is not immediately clear to the caller that this value is scoped to the function and may not be used as a return value.

On second thought, `ROOT_PATH_NS_VAR` probably matches developer expectations, but this is still unclear in non-ARC code.

In summary:

1. The use of these macros are error-prune
2. The theoretical usage of these macros are severely limited. Mostly, programs either have a path they want to reference as a literal, or they get a path from another function which should have produced a valid path in the first place.
3. Writing the correct code using `THEOS_PACKAGE_INSTALL_PREFIX` directly is not challenging
4. I have found 3 instance of these macros being used in existing projects (see below)

From above point 4:
1. https://github.com/CreatureSurvive/TweakSettings/blob/4162605bfebd2a925a681c81ed5f03c3152cb901/TweakSettings-App/main.m#L28 seems like a valid use, however using `THEOS_PACKAGE_INSTALL_PREFIX` directly here would be trivial, as mentioned in previous point 3
2. https://github.com/Traurige/Eneko/blob/3a260f983ac59790665b87bb9c8644d9bdaafa08/Preferences/Controllers/EnekoRootListController.m#L99C1-L100 the macro should be used on the line above which contains the string literal
3. https://github.com/AnthoPakPak/QuickPrefs/blob/ea3d221d3833e8b2e7743bb9fc163a351d692f4d/Prefs/Preferences.m#L3 looks like incorrect usage of the `defaults` subsystem

Checklist
---------
- [x] New code follows the [code rules](https://github.com/theos/headers/blob/master/README.md#code-rules)
- [x] My contribution is code written by myself from reverse-engineered headers, licensed into the Public Domain as per [LICENSE.md](LICENSE.md); or, code written by myself / taken from an existing project, released under an OSI-approved license, and I've added relevant licensing credit to [LICENSE.md](LICENSE.md)

Does this close any currently open issues?
------------------------------------------

No, although it addresses comments made on #81 after the PR was merged.

Where has this been tested?
---------------------------
**Operating System:** macOS

**Platform:** …

**Target Platform:** …

**Toolchain Version:** `Apple clang version 14.0.3`

**SDK Version:** …

### CODE REVIEW

**Comment by leptos-null** [comment]
Copying in @CreatureSurvive: how would you feel about removing this macro given that you use it in your codebase?

**Comment by leptos-null** [comment]
Given that `ROOT_PATH_NS_VAR` _is_ known to be used in existing codebases and that memory management of the retuned value should match developer exceptions, perhaps only `ROOT_PATH_VAR` should be removed, and keep the existing `ROOT_PATH_NS_VAR`.

**Comment by opa334** [comment]
I think the Macro is fine to have with alloca, you could also consider making the variable static, then it works properly all of the time

**Comment by L1ghtmann** [comment]
TLDR: there are valid reasons for both keeping and doing away with the _VAR macros and it seems like more research/consideration may be necessary if we want to definitively say that the current impl is worth doing away with. 

#### Thought dump:
> Given that ROOT_PATH_NS_VAR is known to be used in existing codebases and that memory management of the retuned value should match developer exceptions, perhaps only ROOT_PATH_VAR should be removed, and keep the existing ROOT_PATH_NS_VAR.

- That's a tough call. I'm not sure it's worth keeping one of the _VAR macros and not the other as one might expect one to exist if the other does. 

> Writing the correct code using THEOS_PACKAGE_INSTALL_PREFIX directly is not challenging

- Though it is easy to append a pre-validated path to the install prefix, and I can't imagine that'll be too much of a burden for smaller projects, for larger projects that r/w/x on-device, I can see many resorting to their own macros (as is often done for localization, for example).

> On second thought, ROOT_PATH_NS_VAR probably matches developer expectations, but this is still unclear in non-ARC code.

- Given Theos's defaults, those compiling w/o ARC in 2023 are almost certainly doing so knowingly. Consequently, I'd argue they're cognizant enough to check macro impl before use.

> The theoretical usage of these macros are severely limited. Mostly, programs either have a path they want to reference as a literal, or they get a path from another function which should have produced a valid path in the first place.

- It's logical to expect one to provide properly formatted paths to the macros, but mistakes happen, so having append format validation done automatically is quite valuable, especially considering incorrectly formatted paths can lead to mysterious errors or broken functionality. 
  - I believe someone had an issue with this not too long ago in the Theos discord -- where a missing trailing slash on a file path provided to the non-VAR equivalent macros caused unexpected behavior and a code review was required to diagnose the issue. 
    - Though the fix was as easy as switching to the path-centric macros, this reveals another potential problem: we're relying on the user to choose the correct macro as, prior to the introduction of the documentation you've added here, the functionality of said macros may not be immediately obvious to new devs. Thus if we keep the macros, I think it'd be worth documenting their usage as you've done with the others. 
  - I should mention that, yes, I do realize these last two points clash in that the latter assumes the user will make mistakes and thus provides a fallback correction mechanism while the former assumes the user has a firm grasp of their code and understands the implications of modifying build settings. I guess we'll have to make a call as to what assumption we want to go with.

**Comment by leptos-null** [comment]
> I think the macro is fine to have with alloca, you could also consider making the variable static, then it works properly in all cases

Except multithreading. Even if we did `_Thread_local static` that seems strange to me.

**Comment by leptos-null** [comment]
Thank you L1ghtmann for the thoughts.

> I'm not sure it's worth keeping one of the _VAR macros and not the other as one might expect one to exist if the other does.

I thought this too, but Obj-C is a higher level language, I think it's fair.

> Given Theos's defaults, those compiling w/o ARC in 2023 are almost certainly doing so knowingly. Consequently, I'd argue they're cognizant enough to check macro impl before use.

Very fair. Agree we should keep `NS_VAR` at this point.

**Comment by L1ghtmann** [comment]
In summary, it seems like we have four options:
1) `_Thread_local static`
2) `malloc`
3) require the user to pass a buffer
4) remove it outright

- 1 could lead to issues within the same thread (not sure how big of a concern this is)
- 2 would require the user to manage the memory (may not be immediately apparent) 
  - if we make the assumption, as mentioned before, that users will review the macros' usage before implementation, this could be a plausible solution
- 3 would seemingly solve both issues but deviates from the other macros' behavior 
- 4 seems reasonable if all other options have a point of contention

**Comment by leptos-null** [comment]
I've tested `_Thread_local static` and haven't found any issues. The downside of this technique is that every use of `ROOT_PATH_VAR` would effectively allocate an additional `PATH_MAX` bytes in the binary itself.

**Comment by leptos-null** [comment]
Thank you all for the lengthy and thoughtful discussion on this thread.
I've decided to remove the `ROOT_PATH_VAR` macro completely. The trade-offs of each possible memory allocation method is best selected by the programmer. This PR provides example code of how one may approach this in C.

**Comment by opa334** [comment]
The problem here is that the rootless 2.0 implementation will require doing exactly this for all macros, so we need to figure a good way to handle it now.

**Comment by TheGithubJoshua** [comment]
> The problem here is that the rootless 2.0 implementation (included in Dopamine 2.0 with the main intend being the deprecation of RootHides iphoneos-arm64e) will require doing exactly this for all macros (because the prefix will be gotten at runtime and not hardcoded), so we need to figure a good way to handle it now.

@opa334 what is the possibility that DragonBuild will support Rootless 2.0 when using the Theos makefile?

**Comment by L1ghtmann** [comment]
@TheGithubJoshua This is not the place to ask this. Dragon != Theos.

**Comment by TheGithubJoshua** [comment]
@L1ghtmann couldn't find anywhere else to ask.

**Comment by L1ghtmann** [comment]
So you chose a Theos headers issue....

**Comment by leptos-null** [comment]
We'll make a separate commit for adding the comments and documentation
